### ISSUE TITLE
Parameter order in Swagger document doesn't correspond to API

### ISSUE THREAD

**Issue by ulrichb** [nestjs/swagger#116]
## I'm submitting a...

<pre><code>
[ ] Regression 
[x] Bug report
[ ] Feature request
[ ] Documentation issue or request
[ ] Support request => Please do not submit support request here, instead post your question on Stack Overflow.
</code></pre>

## Current behavior

Controller method:
![image](https://user-images.githubusercontent.com/388796/43281040-d848c810-9112-11e8-879d-888a0f0d92af.png)

Swagger result:

![image](https://user-images.githubusercontent.com/388796/43281406-dd14f4bc-9113-11e8-8a16-f72a563784cd.png)

=> `configurationName` and `module` is reversed :/
This has the consequence that e.g. NSwag generated methods also have the wrong parameter order.

## Expected behavior
The Swagger doc respects the source-code order.

## Bug cause

The problem is that the route parameters decorator metadata (`ROUTE_ARGS_METADATA`) contains a object, not an array (as `DECORATORS.API_PARAMETERS` does). => [`exploreApiReflectedParametersMetadata`](https://github.com/nestjs/swagger/blob/941a82ab3b2f6e7f986d144520937215deda58c8/lib/explorers/api-parameters.explorer.ts#L71) doesn't return the correct order.

See the following debugger screenshot.

![image](https://user-images.githubusercontent.com/388796/43281579-59eb5d8c-9114-11e8-90e6-803cc68eb10e.png)

Sorting `parameters` by the object keys (i.e. `5:0`, `5:1` in the shown example) explicitly, using ...

    const sortedParameters = lodash_1.map(lodash_1.sortBy(lodash_1.map(parameters, (item, name) => ({item, name})), 'name'), x => x.item);

... fixes the issue (see also debugger screenshot).

**Comment by ulrichb** [comment]
**Edit:** This may depend on the node version because the swagger output order depends on the object key order, [which is not guaranteed](https://stackoverflow.com/a/5525820/50890).

My node version: 10.7.0.

**Comment by ulrichb** [comment]
[Background info for the reverse order of the decorator execution.](https://github.com/Microsoft/TypeScript/issues/7258#issuecomment-190897712)

**Comment by edevil** [comment]
@ulrichb I don't know how you manage to avoid using ```ApiImplicitParam```, without it my params don't event appear on the swagger document object. In addition, how can we download the swagger document in json format through the explorer? I want to compare with what you get. Thanks.

**Comment by ulrichb** [comment]
Hi @edevil 
> how can we download the swagger document in json format through the explorer? 

Either use the url `.../swagger-json` (hosted next to the UI version `.../swagger`).

Or you can dump it in your code:
```JS
const swaggerDocument = 
    SwaggerModule.createDocument(app, new DocumentBuilder()./*...*/.build());

console.log(JSON.stringify(swaggerDocument, undefined, 1));
```

> I don't know how you manage to avoid using ApiImplicitParam, without it my params don't event appear on the swagger document object

I don't use `ApiImplicitParam` at all because the swagger generator correctly detects the `@Param`-decorated parameters.

I think the  `@Param`-parameters are extracted [in `exploreApiReflectedParametersMetadata ` using `ROUTE_ARGS_METADATA`](https://github.com/nestjs/swagger/blob/d1791c4b2efccfde5508abd3edffed2fd275eeda/lib/explorers/api-parameters.explorer.ts#L72-L79).

And [here happens the "union"](https://github.com/nestjs/swagger/blob/d1791c4b2efccfde5508abd3edffed2fd275eeda/lib/explorers/api-parameters.explorer.ts#L33-L60) between the two sources.

**Comment by adrianpietka** [comment]
The same problem exists for `@ApiImplicitParam` decorator. Swagger documentation was generated with inverted order for parameters.

My Node.js version: 8.11.3

**Comment by edevil** [comment]
> > how can we download the swagger document in json format through the explorer?
> 
> Either use the url `.../swagger-json` (hosted next to the UI version `.../swagger`),

Thanks!

> I don't use `ApiImplicitParam` at all because the swagger generator correctly detects the `@Param`-decorated parameters.

I know it should be doing that, but it does not happen for me... I have to specify `ApiImplicitParam`. And I have to reverse the order for it to be correct:

```
  @Get(':id/:id2')
  @ApiImplicitParam({ name: 'id2' })
  @ApiImplicitParam({ name: 'id' })
  findOne(@Param('id') id, @Param('id2') id2) {
    return `This action returns a #${id} cat #${id2}`;
  }
```

```
      "get": {
        "parameters": [
          {
            "name": "id",
            "required": true,
            "in": "path",
            "type": ""
          },
          {
            "name": "id2",
            "required": true,
            "in": "path",
            "type": ""
          }
        ],
```

**Comment by ulrichb** [comment]
@edevil Do you use TypeScript? If yes, could you post the generated .JS of `findOne()` (to be able to see the generated decorator metadata)?

**Comment by edevil** [comment]
@ulrichb Is this what you asked for?:
```
__decorate([
    common_1.Get(':id/:id2'),
    swagger_1.ApiImplicitParam({ name: 'id2' }),
    swagger_1.ApiImplicitParam({ name: 'id' }),
    __param(0, common_1.Param('id')), __param(1, common_1.Param('id2')),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, Object]),
    __metadata("design:returntype", void 0)
], CatsController.prototype, "findOne", null);
```

**Comment by ulrichb** [comment]
@edevil Yep. 

Unfortunately I find no difference to my generated decorator code.

As your problem is a another issue (unrelated to the swagger _order_), I think its best to create a complete repro sample and open a new issue.

**Comment by ulrichb** [comment]
Oh, I just saw the `__metadata("design:paramtypes", [Object, Object])` (you have an implicit `any` on your controller parameters) please try to type them _with a scalar type_ (sth. like `@Param('id') id: string`).

**Comment by edevil** [comment]
@ulrichb Ah, that worked, thanks. By specifying a concrete type I don't need to use ```ApiImplicitParam```. And my parameters are in reverse order as "expected".

**Comment by maxfriedmann** [comment]
So whats the status... We're generating clients based on the swagger definition, and we're seeing this issue everywhere.

**Comment by Srso** [comment]
Hi, we migrate to Spring boot 2 and we generate clients based on the swagger definition too, but we have same problem, parameters have other order as in the code :( . Subsequently is client generated with word order of parameters too -> now clients cannot be used. Can you help me, how can the order of parameters be maintained?

**Comment by Srso** [comment]
I used older version 2.8.0 of springfox :(.

**Comment by kamilmysliwiec** [comment]
Fixed in 4.0.0

**Comment by ulrichb** [comment]
@kamilmysliwiec Yipeee!! Many thanks again.
